src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,316 @@
+                    // TODO:  Also check that sequentiality is kept.
+                }
+            });
+
+            return task;
+        }
+
+        /// <summary>
+        /// Obtains a block hash and height of the current tip.
+        /// </summary>
+        /// <returns><see cref="HashHeightPair"/> of current <see cref="ProvenBlockHeader"/> tip.</returns>
+        private HashHeightPair GetTipHashHeight()
+        {
+            HashHeightPair tip = this.blockHashHeightPair;
+
+            using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+            {
+                if (tip == null)
+                {
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    Row<byte[], HashHeightPair> row = transaction.Select<byte[], HashHeightPair>(BlockHashHeightTable, blockHashHeightKey);
+
+                    if (row.Exists)
+                        this.blockHashHeightPair = row.Value;
+                }
+            }
+
+            return this.blockHashHeightPair;
+        }
+
+        /// <summary>
+        /// Set's the hash and height tip of the new <see cref="ProvenBlockHeader"/>.  
+        /// </summary>
+        /// <param name="transaction">Open DBreeze transaction.</param>
+        /// <param name="newTip">Hash height pair of the new block tip.</param>
+        private void SetTip(DBreeze.Transactions.Transaction transaction, HashHeightPair newTip)
+        {
+            Guard.NotNull(newTip, nameof(newTip));
+
+            this.blockHashHeightPair = newTip;
+
+            transaction.Insert<byte[], HashHeightPair>(BlockHashHeightTable, blockHashHeightKey, newTip);
+        }
+
+        /// <summary>
+        /// Inserts <see cref="ProvenBlockHeader"/> items into to the database.
+        /// </summary>
+        /// <param name="transaction">Open DBreeze transaction.</param>
+        /// <param name="headers">List of <see cref="ProvenBlockHeader"/> items to save.</param>
+        /// <param name="newTip">Hash and height of the new repository's tip.</param>
+        private void InsertHeaders(DBreeze.Transactions.Transaction transaction, List<ProvenBlockHeader> headers, HashHeightPair newTip)
+        {
+            int tipHeight = newTip.Height;
+
+            var headerDict = new Dictionary<int, ProvenBlockHeader>();
+
+            // Gather headers.
+            for (int i = tipHeight; i > -1; i--)
+                headerDict[i] = headers[i];
+
+            var sortedHeaders = headerDict.ToList();
+
+            sortedHeaders.Sort((pair1, pair2) => pair1.Key.CompareTo(pair2.Key));
+
+            foreach (KeyValuePair<int, ProvenBlockHeader> header in sortedHeaders)
+                transaction.Insert<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, header.Key.ToBytes(false), header.Value);
+
+            // Store the latest ProvenBlockHeader in memory.
+            this.provenBlockHeaderTip = headers.LastOrDefault();
```

Comment:
Can this be Default? there should always be a header so no need to use default

Reply:
Already mentioned that. 

Let's wait for another round of review, otherwise we will duplicate ideas about fixes. 

Next review is on you 🙂

Reply:
ok